cabal-install-solver/src/Distribution/Solver/Modular/Message.hs

```diff
@@ -206,12 +208,81 @@ data MergedPackageConflict = MergedPackageConflict {
   , versionConflict :: Maybe VR
   }
 
-showQPNPOpt :: QPN -> POption -> String
-showQPNPOpt qpn@(Q _pp pn) (POption i linkedTo) =
+data ProgressAction =
+    Trying
+  | Skipping
+  | Rejecting
+
+instance Show ProgressAction where
+  show Trying = "trying: "
+  show Skipping = "skipping: "
+  show Rejecting = "rejecting: "
+
+showQFNBool :: ProgressAction -> QFN -> Bool -> String
+showQFNBool a q b = show a ++ Flag.showQFNBool q b
+
+showQSNBool :: ProgressAction -> QSN -> Bool -> String
+showQSNBool a q b = show a ++ Flag.showQSNBool q b
+
+showOptions :: ProgressAction -> QPN -> [POption] -> String
+showOptions a q [p] = show a ++ showOption q p
+showOptions a q ps = show a ++ showIsOrVs q (tryVs ps)
+
+showOption :: QPN -> POption -> String
+showOption qpn@(Q _pp pn) (POption i linkedTo) =
   case linkedTo of
     Nothing  -> showPI (PI qpn i) -- Consistent with prior to POption
     Just pp' -> showQPN qpn ++ "~>" ++ showPI (PI (Q pp' pn) i)
 
+-- | A list of versions, or a list of instances.
+data IsOrVs = Is [POption] | Vs [Ver] deriving Show
+
+-- | Try to convert a list of options to a list of versions, or a list of
+-- instances if any of the options is linked (installed).  Singleton lists or
+-- empty lists are always converted to Is.
+-- >>> tryVs [v0, v1]
+-- Vs [mkVersion [0],mkVersion [1]]
```

Comment:
The two functions with doctests I would say are a stable interface, they do contribute to the logged output of the solver. As unit tests, they are smaller units.

I've encountered various problems with doctests over the years but haven't had any problems so far with these in this project. Here's how I run doctests:

```
$ make doctest-install
$ cabal repl --with-ghc=doctest --repl-options="-w" --project-file=cabal.project.doctest cabal-install-solver
```

A bonus of doctests (stating the obvious) is that they also serve as documentation.

I don't think failing doctests could be any easier to modify, the failure report has the actual output. This can be copied from the report and pasted directly as the expectation.

Reply:
I have no problem with using doctests (as long as they are added to the CI), but I think that `tryVs` and `showIsOrVs` shouldn't be tested directly because they don't form a stable interface. I thought about it some more, and the reason that it seems unstable is that it exposes the new `IsOrVs` type, which seems very tightly tied to the current logic for displaying rejected package options. If someone made a further improvement to the format for displaying rejected versions, it is very likely that they would modify or even remove the `IsOrVs` type.

Another idea for testing a more stable interface would be to make a new function that calls both `tryVs` and `showIsOrVs` and add the doctests to it. The function would convert a `QPN` and a `[POption]` to a string, which I think is functionality that would continue to be needed through a wide variety of potential improvements. Another benefit of testing at that level is that when contributors make additional improvements to the format of the rejecting message, they can simply update the expected output for the existing tests (in addition to adding new tests), and the diff will clearly show how the format was improved.

Reply:
I've added doctests in `cabal-install-solver` to the Makefile `doctest` target recipe so they'll run in CI.

We don't export functions `tryVs`, `showIsOrVs` but can test them with doctests the way we have it set up.